edns.c

```diff
@@ -0,0 +1,244 @@
+}
+
+/* note: data must be allocated memory */
+ldns_edns_option *
+ldns_edns_new(ldns_edns_option_code code, size_t size, void *data)
+{
+    ldns_edns_option *edns;
+    edns = LDNS_MALLOC(ldns_edns_option);
+    if (!edns) {
+        return NULL;
+    }
+    ldns_edns_set_size(edns, size);
+    ldns_edns_set_code(edns, code);
+    ldns_edns_set_data(edns, data);
+    return edns;
+}
+
+void
+ldns_edns_deep_free(ldns_edns_option *edns)
+{
+    if (edns) {
+        if (edns->_data) {
+            LDNS_FREE(edns->_data);
+        }
+        LDNS_FREE(edns);
+    }
+}
+
+void 
+ldns_edns_free(ldns_edns_option *edns)
+{
+    if (edns) {
+        LDNS_FREE(edns);
+    }
+}
+
+ldns_edns_option_list*
+ldns_edns_option_list_new()
+{
+    ldns_edns_option_list *options_list = LDNS_MALLOC(ldns_edns_option_list);
+    if(!options_list) {
+        return NULL;
+    }
+
+    options_list->_option_count = 0;
+    options_list->_options = NULL;
+    return options_list;
+}
+
+void
+ldns_edns_option_list_free(ldns_edns_option_list *options_list)
+{
+    if (options_list) {
+        LDNS_FREE(options_list->_options);
+        LDNS_FREE(options_list);
+    }
+}
+
+void
+ldns_edns_option_list_deep_free(ldns_edns_option_list *options_list)
+{
+    size_t i;
+
+    if (options_list) {
+        for (i=0; i < ldns_edns_option_list_get_count(options_list); i++) {
+            ldns_edns_free(ldns_edns_option_list_get_option(options_list, i));
+        }
+        ldns_edns_option_list_free(options_list);
+    }
+}
+
+
+size_t
+ldns_edns_option_list_get_count(const ldns_edns_option_list *options_list)
+{
+    if (options_list) {
+        return options_list->_option_count;
+    } else {
+        return 0;
+    }
+}
+
+void
+ldns_edns_option_list_set_count(ldns_edns_option_list *options_list, size_t count)
+{
+    assert(options_list); // @TODO does this check need to check more?
+    options_list->_option_count = count;
+}
+
+ldns_edns_option *
+ldns_edns_option_list_get_option(const ldns_edns_option_list *options_list, size_t index)
+{
+    if (index < ldns_edns_option_list_get_count(options_list)) {
+        return options_list->_options[index];
+    } else {
+        return NULL;
+    }
+}
+
+ldns_edns_option *
+ldns_edns_option_list_set_option(ldns_edns_option_list *options_list,
+    const ldns_edns_option *option, size_t index)
+{
+    ldns_edns_option* old;
+
+    assert(options_list != NULL);
+
+    if (index < ldns_edns_option_list_get_count(options_list)) {
+        return NULL;
+    }
+
+    if (option == NULL) {
+        return NULL;
+    }
+
+    old = ldns_edns_option_list_get_option(options_list, index);
+
+    /* overwrite the pointer of "old" */
+    options_list->_options[index] = (ldns_edns_option*)option;
+    return old;
+}
+
+bool
+ldns_edns_option_list_push(ldns_edns_option_list *options_list,
+    const ldns_edns_option *option)
+{
+    assert(options_list != NULL);
+
+    if (option != NULL) {
+
+        // @TODO rethink reallocing per push
```

Comment:
Since we're not expecting long lists of EDNS options like in the RR list, this might be superfluous. For example, the RR list _capacity gets initiated at 8, which too large for the number of EDNS options in my opinion. WDYT?